### PR TITLE
planner: push down Lock to PointGet and BatchPointGet

### PR DESCRIPTION
**Pull request by coocood**
### What problem does this PR solve? 
- Lock non-exist keys if DataSource is PointGet or BatchPointGet.
- Make BatchPointGet supports Update and Delete.

### What is changed and how it works?
PointGet and BatchPointGet supports lock by itself, so we can push down
the lock operation down to PointGet and BatchPointGet to lock non-exist keys.

They also read data from MemBuffer, so UnionScan can be eliminated.

By add a function in postOptimize, we can eliminate UnionScan and push down lock to PointGet and BatchPointGet.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Related changes

 - Need to update the documentation

Release note

 - Lock non-exist keys if DataSource is PointGet or BatchPointGet.

### CODE REVIEW

**Comment by coocood** [comment]
@lzmhhh123 @cfzjywxk @tiancaiamao PTAL

File: planner/core/optimizer.go [hunk 1]
```diff
@@ -187,6 +188,72 @@ func physicalOptimize(logic LogicalPlan) (PhysicalPlan, float64, error) {
...
+			pointGet = x
+		case *BatchPointGetPlan:
+			batchPointGet = x
+		case *PhysicalLock:
+			physLock = x
+		case *PhysicalUnionScan:
+			unionScan = x
+		}
+		return true
+	})
```
Comment (lzmhhh123) [inline planner/core/optimizer.go]: How about doing `transformPhysicalPlan` here to reduce plan tree iteration?

File: planner/core/optimizer.go [hunk 2]
```diff
@@ -187,6 +188,72 @@ func physicalOptimize(logic LogicalPlan) (PhysicalPlan, float64, error) {
...
+	if pointGet == nil && batchPointGet == nil {
+		return p
+	}
+	if physLock == nil {
+		return p
```
Comment (lysu) [inline planner/core/optimizer.go]: can we eliminate `unionScan` even if it's a select without `for update`?

```
begin pessimistic;
insert t values (8, 8)
explain select c + 1 from t where k = 2 and c = 2; // no need unionScan?
```

File: session/pessimistic_test.go [hunk 3]
```diff
@@ -927,6 +928,46 @@ func (s *testPessimisticSuite) TestPessimisticReadCommitted(c *C) {
...
+	tk := testkit.NewTestKitWithInit(c, s.store)
+	tk1 := testkit.NewTestKitWithInit(c, s.store)
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t (k int primary key, c int)")
+	tk.MustExec("insert t values (1, 1), (3, 3), (5, 5)")
+
+	// verify that select with project and filter on a non exists key still locks the key.
+	tk.MustExec("begin pessimistic")
+	tk.MustExec("insert t values (8, 8)") // Make the transaction dirty.
+	tk.MustQuery("select c + 1 from t where k = 2 and c = 2 for update").Check(testkit.Rows())
```
Comment (lysu) [inline session/pessimistic_test.go]: if we change this sql to 

`select c + 1 from t where k = (2, 3) and c = 2 for update`

`(3, 3)` also be locked after this pr, but mysql seems do the same things - -

File: planner/core/optimizer.go (hunk 1 above)
Comment (coocood) [reply planner/core/optimizer.go]: transform is post-order, iterate is pre-order.
Use pre-order to iterate so we can stop early, and we can not do transform in pre-order.

File: session/pessimistic_test.go (hunk 3 above)
Comment (coocood) [reply session/pessimistic_test.go]: Yes, it is expected, we lock it even if it doesn't pass the filter.

**Pull request merged**